### PR TITLE
Fix shareGradInput for pre-activation ResNet

### PR DESCRIPTION
**Pull request by colesbury**
To test compare the loss after the first few iterations with:

 th main.lua -netType preresnet -depth 18 -cudnn deterministic -nThreads 1 -manualSeed 1
 th main.lua -netType preresnet -depth 18 -cudnn deterministic -nThreads 1 -manualSeed 1 -shareGradInput true

### CODE REVIEW

**Comment by fmassa** [comment]
What about using `optnet` for the gradient sharing ? @szagoruyko [has a diff which adds it](https://github.com/szagoruyko/fb.resnet.torch/commit/5a7738540a04dd9fc6016633c9106cf48cc79ab3).
That could potentially avoid problems like the one in https://github.com/facebook/fb.resnet.torch/issues/43

**Comment by colesbury** [comment]
I tried Sergey's diff and it didn't seem to work (the training loss changed when running with `-shareGradInput` flag)

**Comment by fmassa** [comment]
I think this difference in the training loss might be due to BatchNorm's `running_mean` and `running_var` changing because of the `forward` calls inside `optnet`.
I [just added gradchecks for the `pre-resnet` architecture](https://github.com/fmassa/optimize-net/commit/29a3731c15f615891938d84bdc8cda02eff800aa) and at least for the cifar10 dataset everything seems fine. Note that I replaced the BatchNorms by `MulConstant` in my tests to avoid the aforementioned effects.
I'll check the gradient checks on imagenet dataset and I'll come back to you.

(On an unrelated note, I was considering adding a `freezeStats` option to BatchNorm, to allow fixing the internal parameters of BatchNorm without having to worry about `training`/`evaluation` mode, this could be handy in this situation).

**Comment by szagoruyko** [comment]
fyi, I am using optnet with clearState and it works very well, eg. checkpoints are 90Mb for resnet18.
